python_modules/libraries/dagster-datahub/dagster_datahub_tests/test_resources.py

```diff
@@ -0,0 +1,108 @@
+from datahub.metadata.schema_classes import ChangeTypeClass, DatasetPropertiesClass
+
+from dagster import DagsterResourceFunctionError, OpExecutionContext, build_op_context, op
+
+
+@responses.activate
+def test_datahub_rest_emitter_resource():
+    @op(required_resource_keys={"datahub"})
+    def datahub_op(context: OpExecutionContext):
+        assert context.resources.datahub is not None
+        dataset_properties = DatasetPropertiesClass(
+            description="This table stored the canonical User profile",
+            customProperties={"governance": "ENABLED"},
+        )
+
+        # Construct a MetadataChangeProposalWrapper object.
+        metadata_event = MetadataChangeProposalWrapper(
+            entityType="dataset",
+            changeType=ChangeTypeClass.UPSERT,
+            entityUrn=builder.make_dataset_urn("bigquery", "my-project.my-dataset.user-table"),
+            aspectName="datasetProperties",
+            aspect=dataset_properties,
+        )
+        context.resources.datahub.emit(metadata_event)
+
+    with responses.RequestsMock() as rsps:
+        # Creating the datahub_rest_emitter resource will run the DatahubRestEmitter.test_connection() method.
+        # The response needs to be mocked out prior to the resource gets instantiated.
+        # This is because creating the resource its the config endpoint as a healthcheck.
+        rsps.add(rsps.GET, "http://foobar:8080/config", status=200, json={"noCode": "true"})
+        rsps.add(rsps.POST, "http://foobar:8080/aspects?action=ingestProposal", status=200, json={})
+        context = build_op_context(
+            resources={
+                "datahub": datahub_rest_emitter.configured({"connection": "http://foobar:8080"})
+            }
+        )
+        datahub_op(context)
+
+
+def test_datahub_emitter_resource_failure():
+    """Assert creating the resource without the mocked response fails to construct.
+    DatahubRestEmitter.test_connection will fail due to the lack of a mock"""
+    with pytest.raises(DagsterResourceFunctionError):
+
+        @op(required_resource_keys={"datahub"})
+        def bad_datahub_op():
+            # Error should be thrown prior to executing the op.
+            # Should Fail if it gets to this point
+            assert False
+
+        context = build_op_context(
+            resources={
+                "datahub": datahub_rest_emitter.configured(
+                    {
+                        "connection": "http://foobar:8080",
+                        "read_timeout_sec": 0.01,
+                        "connect_timeout_sec": 0.01,
+                        "retry_max_times": 1,
+                    }
+                )
+            }
+        )
+        bad_datahub_op(context)
+
+
+def test_datahub_kafka_emitter_resource():
+    @op(required_resource_keys={"datahub"})
+    def datahub_op(context: OpExecutionContext):
+        assert context.resources.datahub is not None
+
+    context = build_op_context(
```

Comment:
Does this need to be mocked out as well?

Getting failures here: https://buildkite.com/dagster/dagster/builds/32538

Reply:
@prha anyway i can access dagster's buildkite to see the error?  Or paste it here?

Reply:
Here's the snippet of the stacktrace, with the full raw log attached:

```
[2022-07-13T16:19:37Z] self = <datahub.emitter.kafka_emitter.DatahubKafkaEmitter object at 0x7f1a686b1370>
--
  | [2022-07-13T16:19:37Z] config = KafkaEmitterConfig(connection=KafkaProducerConnectionConfig(bootstrap='foobar:9092', schema_registry_url='foobar:8081'... topic_routes={'MetadataChangeEvent': 'MetadataChangeEvent_v4', 'MetadataChangeProposal': 'MetadataChangeProposal_v1'})
  | [2022-07-13T16:19:37Z]
  | [2022-07-13T16:19:37Z]     def __init__(self, config: KafkaEmitterConfig):
  | [2022-07-13T16:19:37Z]         self.config = config
  | [2022-07-13T16:19:37Z]         schema_registry_conf = {
  | [2022-07-13T16:19:37Z]             "url": self.config.connection.schema_registry_url,
  | [2022-07-13T16:19:37Z]             **self.config.connection.schema_registry_config,
  | [2022-07-13T16:19:37Z]         }
  | [2022-07-13T16:19:37Z] >       schema_registry_client = SchemaRegistryClient(schema_registry_conf)
  | [2022-07-13T16:19:37Z]
  | [2022-07-13T16:19:37Z] .tox/py39/lib/python3.9/site-packages/datahub/emitter/kafka_emitter.py:67:
  | [2022-07-13T16:19:37Z] _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
  | [2022-07-13T16:19:37Z]
  | [2022-07-13T16:19:37Z] self = <confluent_kafka.schema_registry.schema_registry_client.SchemaRegistryClient object at 0x7f1a686b1730>
  | [2022-07-13T16:19:37Z] conf = {'url': 'foobar:8081'}
  | [2022-07-13T16:19:37Z]
  | [2022-07-13T16:19:37Z]     def __init__(self, conf):
  | [2022-07-13T16:19:37Z] >       self._rest_client = _RestClient(conf)
  | [2022-07-13T16:19:37Z]
  | [2022-07-13T16:19:37Z] .tox/py39/lib/python3.9/site-packages/confluent_kafka/schema_registry/schema_registry_client.py:292:
  | [2022-07-13T16:19:37Z] _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
  | [2022-07-13T16:19:37Z]
  | [2022-07-13T16:19:37Z] self = <confluent_kafka.schema_registry.schema_registry_client._RestClient object at 0x7f1a686b1eb0>
  | [2022-07-13T16:19:37Z] conf = {'url': 'foobar:8081'}
  | [2022-07-13T16:19:37Z]
  | [2022-07-13T16:19:37Z]     def __init__(self, conf):
  | [2022-07-13T16:19:37Z]         self.session = Session()
  | [2022-07-13T16:19:37Z]
  | [2022-07-13T16:19:37Z]         # copy dict to avoid mutating the original
  | [2022-07-13T16:19:37Z]         conf_copy = conf.copy()
  | [2022-07-13T16:19:37Z]
  | [2022-07-13T16:19:37Z]         base_url = conf_copy.pop('url', None)
  | [2022-07-13T16:19:37Z]         if base_url is None:
  | [2022-07-13T16:19:37Z]             raise ValueError("Missing required configuration property url")
  | [2022-07-13T16:19:37Z]         if not isinstance(base_url, string_type):
  | [2022-07-13T16:19:37Z]             raise TypeError("url must be an instance of str, not "
  | [2022-07-13T16:19:37Z]                             + str(type(base_url)))
  | [2022-07-13T16:19:37Z]         if not base_url.startswith('http'):
  | [2022-07-13T16:19:37Z] >           raise ValueError("Invalid url {}".format(base_url))
  | [2022-07-13T16:19:37Z] E           ValueError: Invalid url foobar:8081
  | [2022-07-13T16:19:37Z]
  | [2022-07-13T16:19:37Z] .tox/py39/lib/python3.9/site-packages/confluent_kafka/schema_registry/schema_registry_client.py:75: ValueError
```
[dagster_build_32538_pytest-dagster-datahub-3-dot-9.log](https://github.com/dagster-io/dagster/files/9113516/dagster_build_32538_pytest-dagster-datahub-3-dot-9.log)